<p style="text-align:center">
    <a href="https://skills.network/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkML0101ENSkillsNetwork20718538-2022-01-01" target="_blank">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png" width="200" alt="Skills Network Logo">
    </a>
</p>

<h1 align="center"><font size="5">Final Project: Classification with Python</font></h1>


<h2>Table of Contents</h2>
<div class="alert alert-block alert-info" style="margin-top: 20px">
    <ul>
    <li><a href="https://#Section_1">Instructions</a></li>
    <li><a href="https://#Section_2">About the Data</a></li>
    <li><a href="https://#Section_3">Importing Data </a></li>
    <li><a href="https://#Section_4">Data Preprocessing</a> </li>
    <li><a href="https://#Section_5">One Hot Encoding </a></li>
    <li><a href="https://#Section_6">Train and Test Data Split </a></li>
    <li><a href="https://#Section_7">Train Logistic Regression, KNN, Decision Tree, SVM, and Linear Regression models and return their appropriate accuracy scores</a></li>
</a></li>
</div>
<p>Estimated Time Needed: <strong>180 min</strong></p>
</div>

<hr>


# Instructions


In this notebook, you will  practice all the classification algorithms that we have learned in this course.


Below, is where we are going to use the classification algorithms to create a model based on our training data and evaluate our testing data using evaluation metrics learned in the course.

We will use some of the algorithms taught in the course, specifically:

1. Linear Regression
2. KNN
3. Decision Trees
4. Logistic Regression
5. SVM

We will evaluate our models using:

1.  Accuracy Score
2.  Jaccard Index
3.  F1-Score
4.  LogLoss
5.  Mean Absolute Error
6.  Mean Squared Error
7.  R2-Score

Finally, you will use your models to generate the report at the end.


# About The Dataset


The original source of the data is Australian Government's Bureau of Meteorology and the latest data can be gathered from [http://www.bom.gov.au/climate/dwo/](http://www.bom.gov.au/climate/dwo/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkML0101ENSkillsNetwork20718538-2022-01-01).

The dataset to be used has extra columns like 'RainToday' and our target is 'RainTomorrow', which was gathered from the Rattle at [https://bitbucket.org/kayontoga/rattle/src/master/data/weatherAUS.RData](https://bitbucket.org/kayontoga/rattle/src/master/data/weatherAUS.RData?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkML0101ENSkillsNetwork20718538-2022-01-01)




This dataset contains observations of weather metrics for each day from 2008 to 2017. The **weatherAUS.csv** dataset includes the following fields:

| Field         | Description                                           | Unit            | Type   |
| ------------- | ----------------------------------------------------- | --------------- | ------ |
| Date          | Date of the Observation in YYYY-MM-DD                 | Date            | object |
| Location      | Location of the Observation                           | Location        | object |
| MinTemp       | Minimum temperature                                   | Celsius         | float  |
| MaxTemp       | Maximum temperature                                   | Celsius         | float  |
| Rainfall      | Amount of rainfall                                    | Millimeters     | float  |
| Evaporation   | Amount of evaporation                                 | Millimeters     | float  |
| Sunshine      | Amount of bright sunshine                             | hours           | float  |
| WindGustDir   | Direction of the strongest gust                       | Compass Points  | object |
| WindGustSpeed | Speed of the strongest gust                           | Kilometers/Hour | object |
| WindDir9am    | Wind direction averaged of 10 minutes prior to 9am    | Compass Points  | object |
| WindDir3pm    | Wind direction averaged of 10 minutes prior to 3pm    | Compass Points  | object |
| WindSpeed9am  | Wind speed averaged of 10 minutes prior to 9am        | Kilometers/Hour | float  |
| WindSpeed3pm  | Wind speed averaged of 10 minutes prior to 3pm        | Kilometers/Hour | float  |
| Humidity9am   | Humidity at 9am                                       | Percent         | float  |
| Humidity3pm   | Humidity at 3pm                                       | Percent         | float  |
| Pressure9am   | Atmospheric pressure reduced to mean sea level at 9am | Hectopascal     | float  |
| Pressure3pm   | Atmospheric pressure reduced to mean sea level at 3pm | Hectopascal     | float  |
| Cloud9am      | Fraction of the sky obscured by cloud at 9am          | Eights          | float  |
| Cloud3pm      | Fraction of the sky obscured by cloud at 3pm          | Eights          | float  |
| Temp9am       | Temperature at 9am                                    | Celsius         | float  |
| Temp3pm       | Temperature at 3pm                                    | Celsius         | float  |
| RainToday     | If there was rain today                               | Yes/No          | object |
| RainTomorrow  | If there is rain tomorrow                             | Yes/No          | float  |

Column definitions were gathered from [http://www.bom.gov.au/climate/dwo/IDCJDW0000.shtml](http://www.bom.gov.au/climate/dwo/IDCJDW0000.shtml?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkML0101ENSkillsNetwork20718538-2022-01-01)



## **Import the required libraries**


In [45]:
# All Libraries required for this lab are listed below. The libraries pre-installed on Skills Network Labs are commented.
#!pip install pandas==1.3.4
#!pip install numpy==1.21.4
#!pip install seaborn==0.9.0
#!pip install matplotlib==3.5.0
#!pip install scikit-learn==0.20.1

!pip install pandas
!pip install numpy
!pip install seaborn
!pip install matplotlib
!pip install scikit-learn
# Note: If your environment doesn't support "!mamba install", use "!pip install"

In [46]:
# Surpress warnings:
def warn(*args, **kwargs):
    pass
import warnings
warnings.warn = warn

In [47]:
import pandas as pd
from sklearn.linear_model import LogisticRegression
from sklearn.linear_model import LinearRegression
from sklearn import preprocessing
import numpy as np
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn import svm
from sklearn.metrics import jaccard_score
from sklearn.metrics import f1_score
from sklearn.metrics import log_loss
from sklearn.metrics import confusion_matrix, accuracy_score
import sklearn.metrics as metrics

### Importing the Dataset


In [48]:
filepath='https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-ML0101EN-SkillUp/labs/ML-FinalAssignment/Weather_Data.csv'

In [49]:
df = pd.read_csv(filepath)

> Note: This version of the lab is designed for JupyterLite, which necessitates downloading the dataset to the interface. However, when working with the downloaded version of this notebook on your local machines (Jupyter Anaconda), you can simply **skip the steps above of "Importing the Dataset"** and use the URL directly in the `pandas.read_csv()` function. You can uncomment and run the statements in the cell below.


In [50]:
#filepath = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-ML0101EN-SkillUp/labs/ML-FinalAssignment/Weather_Data.csv"
#df = pd.read_csv(filepath)

In [51]:
df.head()

Date  MinTemp  MaxTemp  Rainfall  Evaporation  Sunshine WindGustDir  \
0  2/1/2008     19.5     22.4      15.6          6.2       0.0           W   
1  2/2/2008     19.5     25.6       6.0          3.4       2.7           W   
2  2/3/2008     21.6     24.5       6.6          2.4       0.1           W   
3  2/4/2008     20.2     22.8      18.8          2.2       0.0           W   
4  2/5/2008     19.7     25.7      77.4          4.8       0.0           W   

   WindGustSpeed WindDir9am WindDir3pm  ...  Humidity9am  Humidity3pm  \
0             41          S        SSW  ...           92           84   
1             41          W          E  ...           83           73   
2             41        ESE        ESE  ...           88           86   
3             41        NNE          E  ...           83           90   
4             41        NNE          W  ...           88           74   

   Pressure9am  Pressure3pm  Cloud9am  Cloud3pm  Temp9am  Temp3pm  RainToday  \
0       1017.6       1017.4         8         8     20.7     20.9        Yes   
1       1017.9       1016.4         7         7     22.4     24.8        Yes   
2       1016.7       1015.6         7         8     23.5     23.0        Yes   
3       1014.2       1011.8         8         8     21.4     20.9        Yes   
4       1008.3       1004.8         8         8     22.5     25.5        Yes   

   RainTomorrow  
0           Yes  
1           Yes  
2           Yes  
3           Yes  
4           Yes  

[5 rows x 22 columns]

### Data Preprocessing


#### One Hot Encoding


First, we need to perform one hot encoding to convert categorical variables to binary variables.


In [52]:
df_sydney_processed = pd.get_dummies(data=df, columns=['RainToday', 'WindGustDir', 'WindDir9am', 'WindDir3pm'])

Next, we replace the values of the 'RainTomorrow' column changing them from a categorical column to a binary column. We do not use the `get_dummies` method because we would end up with two columns for 'RainTomorrow' and we do not want, since 'RainTomorrow' is our target.


In [53]:
df_sydney_processed.replace(['No', 'Yes'], [0,1], inplace=True)

### Training Data and Test Data


Now, we set our 'features' or x values and our Y or target variable.


In [54]:
df_sydney_processed.drop('Date',axis=1,inplace=True)

In [55]:
df_sydney_processed = df_sydney_processed.astype(float)

In [56]:
features = df_sydney_processed.drop(columns='RainTomorrow', axis=1)
Y = df_sydney_processed['RainTomorrow']

In [57]:
features.head()

MinTemp  MaxTemp  Rainfall  Evaporation  Sunshine  WindGustSpeed  \
0     19.5     22.4      15.6          6.2       0.0           41.0   
1     19.5     25.6       6.0          3.4       2.7           41.0   
2     21.6     24.5       6.6          2.4       0.1           41.0   
3     20.2     22.8      18.8          2.2       0.0           41.0   
4     19.7     25.7      77.4          4.8       0.0           41.0   

   WindSpeed9am  WindSpeed3pm  Humidity9am  Humidity3pm  ...  WindDir3pm_NNW  \
0          17.0          20.0         92.0         84.0  ...             0.0   
1           9.0          13.0         83.0         73.0  ...             0.0   
2          17.0           2.0         88.0         86.0  ...             0.0   
3          22.0          20.0         83.0         90.0  ...             0.0   
4          11.0           6.0         88.0         74.0  ...             0.0   

   WindDir3pm_NW  WindDir3pm_S  WindDir3pm_SE  WindDir3pm_SSE  WindDir3pm_SSW  \
0            0.0           0.0            0.0             0.0             1.0   
1            0.0           0.0            0.0             0.0             0.0   
2            0.0           0.0            0.0             0.0             0.0   
3            0.0           0.0            0.0             0.0             0.0   
4            0.0           0.0            0.0             0.0             0.0   

   WindDir3pm_SW  WindDir3pm_W  WindDir3pm_WNW  WindDir3pm_WSW  
0            0.0           0.0             0.0             0.0  
1            0.0           0.0             0.0             0.0  
2            0.0           0.0             0.0             0.0  
3            0.0           0.0             0.0             0.0  
4            0.0           1.0             0.0             0.0  

[5 rows x 66 columns]

In [58]:
Y.head()

0    1.0
1    1.0
2    1.0
3    1.0
4    1.0
Name: RainTomorrow, dtype: float64

In [59]:
#check the shape of the input data X and Y
print(features.shape)
print(Y.shape)

(3271, 66)
(3271,)


### Linear Regression


#### Q1) Use the `train_test_split` function to split the `features` and `Y` dataframes with a `test_size` of `0.2` and the `random_state` set to `10`.


In [60]:
#Enter Your Code and Execute
x_train, x_test, y_train, y_test = train_test_split(features, Y, test_size=0.2, random_state=10)


#### Q2) Create and train a Linear Regression model called LinearReg using the training data (`x_train`, `y_train`).


In [61]:
#Enter Your Code and Execute
LinearReg = LinearRegression()
LinearReg = LinearReg.fit(x_train, y_train)

#### Q3) Now use the `predict` method on the testing data (`x_test`) and save it to the array `predictions`.


In [62]:
#Enter Your Code and Execute
predictions = LinearReg.predict(x_test)

#### Q4) Using the `predictions` and the `y_test` dataframe calculate the value for each metric using the appropriate function.


In [65]:
#Enter Your Code and Execute
LinearRegression_MAE = metrics.mean_absolute_error(y_test, predictions)
LinearRegression_MSE = metrics.mean_squared_error(y_test, predictions)
LinearRegression_R2 = metrics.r2_score(y_test, predictions)
#LinearR_Accuracy_Score = metrics.accuracy_score(y_test, predictions)
#LinearR_JaccardIndex =  jaccard_score(y_test,predictions)
#LinearR_F1_Score = f1_score(y_test, predictions)

#### Q5) Show the MAE, MSE, and R2 in a tabular format using data frame for the linear model.


In [67]:
#Enter Your Code and Execute
# Create a dictionary with the metrics
data = {
    'Metric': ['MAE', 'MSE', 'R2'],
    'Value': [LinearRegression_MAE, LinearRegression_MSE, LinearRegression_R2]
}

# Create a pandas DataFrame from the dictionary
Report = pd.DataFrame(data)
print(Report)

  Metric     Value
0    MAE  0.256318
1    MSE  0.115721
2     R2  0.427132


### KNN


#### Q6) Create and train a KNN model called KNN using the training data (`x_train`, `y_train`) with the `n_neighbors` parameter set to `4`.


In [68]:
#Enter Your Code and Execute
k = 4
#Train Model and Predict
KNN = KNeighborsClassifier(n_neighbors = k).fit(x_train,y_train)

#### Q7) Now use the `predict` method on the testing data (`x_test`) and save it to the array `predictions`.


In [69]:
#Enter Your Code and Execute
predictions = KNN.predict(x_test)

#### Q8) Using the `predictions` and the `y_test` dataframe calculate the value for each metric using the appropriate function.


In [70]:
#Enter Your Code and Execute
KNN_Accuracy_Score = metrics.accuracy_score(y_test, predictions)
KNN_JaccardIndex =  jaccard_score(y_test,predictions)
KNN_F1_Score = f1_score(y_test, predictions)

In [71]:
# Create a dictionary with the metrics
data = {
    'Metric': ['Accuracy Score', 'Jaccard Index', 'F1 Score'],
    'Value': [KNN_Accuracy_Score, KNN_JaccardIndex, KNN_F1_Score]
}

# Create a pandas DataFrame from the dictionary
Report = pd.DataFrame(data)
print(Report)

           Metric     Value
0  Accuracy Score  0.818321
1   Jaccard Index  0.425121
2        F1 Score  0.596610


### Decision Tree


#### Q9) Create and train a Decision Tree model called Tree using the training data (`x_train`, `y_train`).


In [72]:
#Enter Your Code and Execute
Tree = DecisionTreeClassifier(criterion="entropy", max_depth = 4)
Tree.fit(x_train, y_train)

DecisionTreeClassifier(criterion='entropy', max_depth=4)

#### Q10) Now use the `predict` method on the testing data (`x_test`) and save it to the array `predictions`.


In [73]:
#Enter Your Code and Execute
predictions = Tree.predict(x_test)

#### Q11) Using the `predictions` and the `y_test` dataframe calculate the value for each metric using the appropriate function.


In [74]:
#Enter Your Code and Execute
Tree_Accuracy_Score = metrics.accuracy_score(y_test, predictions)
Tree_JaccardIndex = jaccard_score(y_test, predictions)
Tree_F1_Score = f1_score(y_test, predictions)

In [75]:
# Create a dictionary with the metrics
data = {
    'Metric': ['Accuracy Score', 'Jaccard Index', 'F1 Score'],
    'Value': [Tree_Accuracy_Score, Tree_JaccardIndex, Tree_F1_Score]
}

# Create a pandas DataFrame from the dictionary
Report = pd.DataFrame(data)
print(Report)

           Metric     Value
0  Accuracy Score  0.818321
1   Jaccard Index  0.480349
2        F1 Score  0.648968


### Logistic Regression


#### Q12) Use the `train_test_split` function to split the `features` and `Y` dataframes with a `test_size` of `0.2` and the `random_state` set to `1`.


In [ ]:
#Enter Your Code and Execute

In [76]:
x_train, x_test, y_train, y_test = train_test_split(features, Y, test_size=0.2, random_state=1)

#### Q13) Create and train a LogisticRegression model called LR using the training data (`x_train`, `y_train`) with the `solver` parameter set to `liblinear`.


In [ ]:
#Enter Your Code and Execute

In [77]:
LR = LogisticRegression(solver='liblinear',C=0.01)
LR.fit(x_train, y_train)

LogisticRegression(C=0.01, solver='liblinear')

#### Q14) Now, use the `predict` and `predict_proba` methods on the testing data (`x_test`) and save it as 2 arrays `predictions` and `predict_proba`.


In [ ]:
#Enter Your Code and Execute

In [78]:
predictions = LR.predict(x_test)
predictions

array([0., 0., 1., 0., 0., 1., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 1.,
       0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 1.,
       1., 1., 1., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1.,
       0., 1., 0., 1., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 0., 1.,
       0., 0., 0., 1., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 1.,
       0., 1., 1., 0., 0., 0., 0., 0., 0., 0., 1., 1., 0., 0., 1., 0., 0.,
       0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 0.,
       0., 0., 0., 1., 0., 1., 0., 0., 0., 1., 0., 1., 1., 0., 0., 0., 1.,
       0., 0., 1., 0., 0., 1., 0., 1., 0., 1., 1., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 1., 0., 0., 0., 0., 0., 1., 1., 0., 0., 1., 1., 1., 0.,
       0., 0., 0., 0., 0., 1., 0., 0., 1., 0., 1., 0., 0., 0., 0., 0., 0.,
       0., 1., 1., 0., 0., 1., 0., 0., 0., 1., 0., 0., 0., 1., 0., 0., 0.,
       1., 0., 1., 1., 0.

In [79]:
predict_proba = LR.predict_proba(x_test)
predict_proba

array([[0.71865869, 0.28134131],
       [0.97588303, 0.02411697],
       [0.497013  , 0.502987  ],
       ...,
       [0.97264175, 0.02735825],
       [0.73341123, 0.26658877],
       [0.38535874, 0.61464126]])

#### Q15) Using the `predictions`, `predict_proba` and the `y_test` dataframe calculate the value for each metric using the appropriate function.


In [ ]:
#Enter Your Code and Execute

In [80]:
LR_Accuracy_Score = metrics.accuracy_score(y_test, predictions)
LR_JaccardIndex = jaccard_score(y_test,predictions, pos_label=1)
LR_F1_Score = f1_score(y_test, predictions,pos_label=1)
LR_Log_Loss = log_loss(y_test, predict_proba)

In [81]:
# Create a dictionary with the metrics
data = {
    'Metric': ['Accuracy Score', 'Jaccard Index', 'F1 Score','Log Loss'],
    'Value': [LR_Accuracy_Score, LR_JaccardIndex, LR_F1_Score,LR_Log_Loss]
}

# Create a pandas DataFrame from the dictionary
Report = pd.DataFrame(data)
print(Report)

           Metric     Value
0  Accuracy Score  0.827481
1   Jaccard Index  0.484018
2        F1 Score  0.652308
3        Log Loss  0.380085


### SVM


#### Q16) Create and train a SVM model called SVM using the training data (`x_train`, `y_train`).


In [ ]:
#Enter Your Code and Execute

In [82]:
SVM = svm.SVC(kernel='rbf')
SVM.fit(x_train, y_train)

SVC()

#### Q17) Now use the `predict` method on the testing data (`x_test`) and save it to the array `predictions`.


In [ ]:
#Enter Your Code and Execute

In [83]:
predictions = SVM.predict(x_test)

#### Q18) Using the `predictions` and the `y_test` dataframe calculate the value for each metric using the appropriate function.


In [84]:
SVM_Accuracy_Score = metrics.accuracy_score(y_test, predictions)
SVM_JaccardIndex = jaccard_score(y_test,predictions)
SVM_F1_Score = f1_score(y_test, predictions)

In [85]:
# Create a dictionary with the metrics
data = {
    'Metric': ['Accuracy Score', 'Jaccard Index', 'F1 Score'],
    'Value': [SVM_Accuracy_Score, SVM_JaccardIndex, SVM_F1_Score]
}

# Create a pandas DataFrame from the dictionary
Report = pd.DataFrame(data)
print(Report)

           Metric     Value
0  Accuracy Score  0.722137
1   Jaccard Index  0.000000
2        F1 Score  0.000000


### Report


#### Q19) Show the Accuracy,Jaccard Index,F1-Score and LogLoss in a tabular format using data frame for all of the above models.

\*LogLoss is only for Logistic Regression Model


In [86]:
# Define the data as a 2D list
data = [

    [KNN_Accuracy_Score, KNN_JaccardIndex, KNN_F1_Score,"NA"],  # Values for KNN
    [Tree_Accuracy_Score, Tree_JaccardIndex, Tree_F1_Score,"NA"],   # Values for Decition Trees
    [LR_Accuracy_Score, LR_JaccardIndex, LR_F1_Score,LR_Log_Loss],   # Values for Logistic Regression
    [SVM_Accuracy_Score, SVM_JaccardIndex, SVM_F1_Score,"NA"],   # Values for SVM
]

# Create the DataFrame
report = pd.DataFrame(data, columns=['Acurracy','Jacard Index','F1 Score','Log Loss'], index=['KNN', 'Decition Trees','Logistic Regression','SVM'])

print(report)

                     Acurracy  Jacard Index  F1 Score  Log Loss
KNN                  0.818321      0.425121  0.596610        NA
Decition Trees       0.818321      0.480349  0.648968        NA
Logistic Regression  0.827481      0.484018  0.652308  0.380085
SVM                  0.722137      0.000000  0.000000        NA


# Analysis of Results

Let's break down the performance of each model based on the provided metrics: **Accuracy**, **Jaccard Index**, **F1 Score**, and (for Logistic Regression) **Log Loss**.

---

## **1. K-Nearest Neighbors (KNN)**
- **Accuracy**: 0.818 — KNN performs well in terms of classification accuracy, indicating it correctly classifies ~82% of the instances.
- **Jaccard Index**: 0.425 — The Jaccard Index, measuring the similarity between predicted and actual labels, is moderate, suggesting room for improvement in overlapping predictions.
- **F1 Score**: 0.597 — This moderate value indicates that precision and recall are reasonably balanced but not optimal.
- **Log Loss**: Not Applicable (NA) — KNN does not provide probabilistic outputs, so Log Loss isn't computed.

**Comment**: KNN shows good accuracy but struggles with overlap and balancing precision and recall. It might benefit from parameter tuning (e.g., adjusting the number of neighbors).

---

## **2. Decision Trees**
- **Accuracy**: 0.818 — Same as KNN, showing comparable accuracy in classification.
- **Jaccard Index**: 0.480 — Higher than KNN, meaning Decision Trees produce predictions that better match the actual labels.
- **F1 Score**: 0.649 — The best among all models so far, indicating a good balance of precision and recall.
- **Log Loss**: NA — Decision Trees don't output probabilities in a manner suitable for Log Loss.

**Comment**: Decision Trees outperform KNN in the Jaccard Index and F1 Score. They seem to capture the structure of the data better. However, they may be prone to overfitting, so regularization (e.g., pruning) should be considered.

---

## **3. Logistic Regression**
- **Accuracy**: 0.827 — The best accuracy among all models, suggesting it generalizes well to the data.
- **Jaccard Index**: 0.484 — Slightly better than Decision Trees, showing good alignment with actual labels.
- **F1 Score**: 0.652 — Marginally better than Decision Trees, indicating good precision and recall balance.
- **Log Loss**: 0.380 — This low value indicates that Logistic Regression provides well-calibrated probabilities.

**Comment**: Logistic Regression is the best-performing model overall, with the highest accuracy, Jaccard Index, and F1 Score, as well as a low Log Loss. It is likely a good candidate for deployment.

---

## **4. Support Vector Machines (SVM)**
- **Accuracy**: 0.722 — The lowest accuracy among all models, suggesting poor generalization to the data.
- **Jaccard Index**: 0.000 — Indicates no overlap between predicted and actual labels.
- **F1 Score**: 0.000 — Indicates the model failed to classify any instances correctly in terms of true positives.
- **Log Loss**: NA — Log Loss isn't provided for SVM.

**Comment**: SVM performs poorly, likely due to inappropriate parameter selection (e.g., kernel type or regularization strength) or an inability to separate the classes effectively given the feature space.

---

## **Overall Recommendations**
1. **Best Model**: Logistic Regression shows the most consistent and reliable performance across all metrics. It should be prioritized for further tuning or deployment.
2. **Secondary Option**: Decision Trees perform almost as well as Logistic Regression and might be preferable if interpretability is crucial.
3. **KNN**: While accurate, it struggles with precision and recall balance. Parameter tuning or feature scaling might improve its performance.
4. **SVM**: This model underperformed significantly and might need further investigation into its parameterization (e.g., kernel choice, regularization).

If possible, conduct hyperparameter tuning (e.g., grid search or random search) and validate with cross-validation to further refine the models.















<h2 id="Section_5">  How to submit </h2>

<p>Once you complete your notebook you will have to share it. You can download the notebook by navigating to "File" and clicking on "Download" button.

<p>This will save the (.ipynb) file on your computer. Once saved, you can upload this file in the "My Submission" tab, of the "Peer-graded Assignment" section.  


<h2>About the Authors:</h2>

<a href="https://www.linkedin.com/in/joseph-s-50398b136/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkML0101ENSkillsNetwork20718538-2022-01-01">Joseph Santarcangelo</a> has a PhD in Electrical Engineering, his research focused on using machine learning, signal processing, and computer vision to determine how videos impact human cognition. Joseph has been working for IBM since he completed his PhD.

### Other Contributors

[Svitlana Kramar](https://www.linkedin.com/in/svitlana-kramar/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMML0232ENSkillsNetwork30654641-2022-01-01)


## <h3 align="center"> © IBM Corporation 2020. All rights reserved. <h3/>

<!--
## Change Log

| Date (YYYY-MM-DD) | Version | Changed By    | Change Description          |
| ----------------- | ------- | ------------- | --------------------------- |
| 2022-06-22        | 2.0     | Svitlana K.   | Deleted GridSearch and Mock |
--!>